### PR TITLE
Add tryGet/UpdSocket and tryGet/UpdOutput to the Component interface

### PR DESCRIPTION
**Pull request by adamkewley**
Fixes having to use a `try {} catch (...) {}`  to find a socket/output on a Component.

### Brief summary of changes

- Added `tryGetSocket`, `tryUpdSocket`, `tryGetOutput`, and `tryUpdOutput` to the `Component` interface
- `try`, because they will return `nullptr` if the socket/output cannot be found on the component
- There doesn't (appear to be?) a way of doing this without `try...catch` or allocating memory (e.g. with `getOutputs`) and searching in downstream code

### Testing I've completed

- Added basic tests
- Added methods
- Refactored existing methods to use the new ones
- If the tests pass in CI, then it should be fine

### CHANGELOG.md (choose one)

- updated.

- - -
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3673)

### CODE REVIEW

**Review by nickbianco — APPROVED** [review]
LGTM. One minor suggestion for the changelog, but not blocking.

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3673#-:-NoJCxVn7-W6b8_YUIZL:b66pt6f)* status: all files reviewed, 1 unresolved discussion (waiting on @adamkewley)

___
*[`CHANGELOG.md` line 10 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/3673#-NoJC1UM-818U10tUBTF:-NoJC1UM-818U10tUBTG:bfzp1e6) ([raw file](https://github.com/opensim-org/opensim-core/blob/8cf776b112ea9ea1ac36e3fc6a3ff1d42eabd5ab/CHANGELOG.md#L10)):*
> ```Markdown
> 
> Upcoming Release
> ================
> ```

The upcoming release will be `v4.6`. You can update that here or we can handle it in a later merge, either is fine with me.

**Comment by adamkewley** [comment]
Note: can't merge this until I ensure the manual merge commit was ok: there are merge conflicts related to adding list sockets to parts of the API that this PR also touches (line-wise: functionally, not).

**Pull request merged**
